internal/linker/link_adapter.bzl

```diff
@@ -0,0 +1,97 @@
+# Copyright 2017 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""link_adapter is used to add a LinkablePackageInfo to a target that doens't one
+or modify the package_path of a which provides LinkablePackageInfo.
+
+The rule will not allow you to change the package_name of target that already
+provides a LinkablePackageInfo. If a target already provides
+"""
+
+load(
+    "//:providers.bzl",
+    "DeclarationInfo",
+    "ExternalNpmPackageInfo",
+    "JSModuleInfo",
+    "JSNamedModuleInfo",
+    "LinkablePackageInfo",
+)
+
+_ATTRS = {
+    "deps": attr.label_list(
+        mandatory = True,
+    ),
+    "package_name": attr.string(
+        mandatory = True,
+    ),
+    "package_path": attr.string(),
+}
+
+def _impl(ctx):
+    if len(ctx.attr.deps) != 1:
+        fail("Expected a single linkable dependency")
+
+    link_target = ctx.attr.deps[0]
+
+    if ExternalNpmPackageInfo in link_target:
+        fail("3rd party linkable dependency with ExternalNpmPackageInfo not adaptable")
+
+    if LinkablePackageInfo in link_target:
+        if link_target[LinkablePackageInfo].package_name != ctx.attr.package_name:
+            fail("Expected linkable dependency package_name to match adapter package_name")
+
+    providers = [
+        LinkablePackageInfo(
+            package_name = ctx.attr.package_name,
+            package_path = ctx.attr.package_path,
+            path = link_target[LinkablePackageInfo].path,
+            files = link_target[LinkablePackageInfo].files,
+        ),
+    ]
+
+    # Forward other providers that linkable targets such as link_target may provide
+    # See js_library rule which provides all of these for example.
+    if DeclarationInfo in link_target:
```

Comment:
I like that, although it is not immediately obvious how to do it since the providers of a target are not iterable AFAICT
https://docs.bazel.build/versions/master/skylark/lib/Target.html.

```
for p in link_target:
```
fails with
```
ERROR: /private/var/tmp/_bazel_gregmagolan/d0debf76e2dc6a317c7dbd1cd5ae4701/external/npm/@test_multi_linker/lib-b/BUILD.bazel:7:13: in _link_adapter rule @npm//@test_multi_linker/lib-b:lib-b: 
Traceback (most recent call last):
	File "/Users/gregmagolan/oss/rules_nodejs/internal/linker/link_adapter.bzl", line 116, column 25, in _impl
		providers.append(DefaultInfo(
Error: type 'Target' is not iterable
```

```
for k, v in link_target.items():
```
fails with 
```
RROR: /private/var/tmp/_bazel_gregmagolan/d0debf76e2dc6a317c7dbd1cd5ae4701/external/internal_test_multi_linker_sub_deps/@test_multi_linker/lib-a/BUILD.bazel:7:13: in _link_adapter rule @internal_test_multi_linker_sub_deps//@test_multi_linker/lib-a:lib-a: 
Traceback (most recent call last):
	File "/Users/gregmagolan/oss/rules_nodejs/internal/linker/link_adapter.bzl", line 126, column 28, in _impl
		for k, v in link_target.items():
Error: <target //internal/linker/test/multi_linker/lib_a:lib_a> (rule '_js_library') doesn't have provider 'items'
```